Linux perf & allocation flame graph guides (#35)
swift-server/guides: docs/linux-perf.md

# Linux `perf`

## `perf`, what's that?

> The Linux [`perf` tool](https://perf.wiki.kernel.org/index.php/Main_Page) is an incredibly powerful tool, that can amongst other things be used for

Reviewer: ```suggestion
The Linux [`perf` tool](https://perf.wiki.kernel.org/index.php/Main_Page) is an incredibly powerful tool, that can amongst other things be used for:
```

- Sampling CPU-bound processes (or the whole) system to analyse which part of your application is consuming the CPU time
- Accessing CPU performance counters (PMU)
- "user probes" (uprobes) which trigger for example when a certain function in your application runs

> In general, `perf` can count and/or record the call stacks of your threads when a certain event occurs. These events can be triggered by

Reviewer: ```suggestion
In general, `perf` can count and/or record the call stacks of your threads when a certain event occurs. These events can be triggered by:
```

> - Time (eg. 1000 times per second), useful for time profiling. For an example use, see the [CPU performance debugging guide](performance.md).

Reviewer: ```suggestion
- Time (e.g. 1000 times per second), useful for time profiling. For an example use, see the [CPU performance debugging guide](performance.md).
```

- System calls, useful to see where your system calls are happening.
- Various system events, for example if you'd like to know when context switches occur.
- CPU performance counters, useful if your performance issues can be traced down to micro-architectural details of your CPU (such as branch mispredictions). For an example, see [SwiftNIO's Advanced Performance Analysis guide](https://github.com/apple/swift-nio/blob/main/docs/advanced-performance-analysis.md).
- and much more

## Getting `perf` to work

> Unfortunately, how to get `perf` to work depends on your environment. Below, please find a selection of environments and how to get `perf` to work there.

Reviewer: ```suggestion
Unfortunately, getting `perf` to work depends on your environment. Below, please find a selection of environments and how to get `perf` to work there.
```
